core/silkworm/consensus/clique/clique_snapshot.hpp

```diff
@@ -0,0 +1,115 @@
+    std::vector<evmc::address> voters; // Voters for current proposal
+};
+
+// Voting Snapshot for Clique
+class CliqueSnapshot {
+    public:
+        CliqueSnapshot() = default;
+        CliqueSnapshot(uint64_t block_number, evmc::bytes32 hash, std::vector<evmc::address> signers,
+                       std::deque<evmc::address> recents):
+                        block_number_{block_number}, hash_{hash}, signers_{signers}, recents_{recents} {}
+
+        //! \brief Updated snapshot by adding headers
+        //! \param headers: list of headers to add.
+        //! \param config: clique config.
+        ValidationResult add_header(const BlockHeader& header, const evmc::address& signer, const CliqueConfig& config);
+        //! \brief Verify seal for header
+        //! \param header: header to verify.
+        ValidationResult verify_seal(BlockHeader header, const evmc::address& signer);
+
+        //! \brief Checks for authority
+        //! \param block_number: Block to check.
+        //! \param address: Address to check.
+        //! \return if a signer at a given block height is in charge or not.
+        bool is_authority(uint64_t block_number, evmc::address address) const noexcept;
+
+        //! \brief Getter method for signers_.
+        //! \return Snapshot's signers.
+        const std::vector<evmc::address>& get_signers() const noexcept;
+
+        //! \brief Getter method for block_number_.
+        //! \return Snapshot's block number.
+        const uint64_t& get_block_number() const noexcept;
+
+        //! \brief Getter method for hash_.
+        //! \return Snapshot's hash.
+        const evmc::bytes32& get_hash() const noexcept;
+        //! \brief Convert the snapshot in JSON.
+        //! \return The resulting JSON.
+        Bytes to_bytes() const noexcept;
+        //! \brief Decode snapshot from json format.
+        //! \return Decoded snapshot.
+        static CliqueSnapshot from_bytes(ByteView& b, uint64_t& block_number, const evmc::bytes32& hash) noexcept;
+
+    private:
+        // is_vote_valid returns whether it makes sense to cast the specified vote in the
+        // given snapshot context (e.g. don't try to add an already authorized signer).
+        bool is_vote_valid(const evmc::address& address, bool authorize) const noexcept;
+        // cast adds a new vote into the tally.
+        void cast(const evmc::address& address, const evmc::address& signer, bool authorize);
+        // uncast removes a previously cast vote from the tally.
+        void uncast(const evmc::address& address, const evmc::address& signer);
+        // uncast_all removes every cast vote from the signer.
+        void uncast_all(const evmc::address& signer);
+        // update update snapshots and cleans it up from db corruption, etc..
+        void update(const uint64_t& block_number, const evmc::bytes32& hash);
+
+        uint64_t block_number_;                             // Block number where the snapshot was created
```

Comment:
Please default-initialize `block_number_` to 0 to avoid potentially unitialized value.